website/static/js/apiApplication.js

```diff
@@ -0,0 +1,433 @@
+        }.bind(this));
+
+        request.fail(function (xhr, status, error) {
+            $osf.growl('Error',
+                       language.apiOauth2Application.dataSendError,
+                       'danger');
+
+            Raven.captureMessage('Error registering new OAuth2 application', {
+                url: this.apiDetailUrl,
+                status: status,
+                error: error
+            });
+        }.bind(this));
+    },
+    submit: function () {
+        // Validate and dispatch form to correct handler based on view type
+        if (!this.appData().isValid()) {
+            // Turn on display of validation messages
+            this.showMessages(true);
+        } else {
+            this.showMessages(false);
+            if (this.isCreateView()) {
+                this.createApplication();
+            } else {
+                this.updateApplication();
+            }
+        }
+    },
+    deleteApplication: function () {
+        var appData = this.appData();
+        bootbox.confirm({
+            title: 'Deactivate application?',
+            message: language.apiOauth2Application.deactivateConfirm,
+            callback: function (confirmed) {
+                if (confirmed) {
+                    var request = this.client.deleteOne(appData );
+                    request.done(function () {
+                        this.allowExit(true);
+                        // Don't let user go back to a deleted application page
+                        historyjs.replaceState({}, '', this.webListUrl);
+                        this.visitList();
+                    }.bind(this));
+                    request.fail(function () {
+                            $osf.growl('Error',
+                                       language.apiOauth2Application.deactivateError,
+                                       'danger');
+                    }.bind(this));
+                }
+            }.bind(this),
+            buttons:{
+                confirm:{
+                    label:'Deactivate',
+                    className:'btn-danger'
+                }
+            }
+        });
+    },
+    visitList: function () {
+        window.location = this.webListUrl;
+    },
+    cancelChange: function () {
+        if (!this.dirty()) {
+            this.visitList();
+        } else {
+            bootbox.confirm({
+                title: 'Discard changes?',
+                message: language.apiOauth2Application.discardUnchanged,
+                callback: function(confirmed) {
+                    if (confirmed) {
+                        this.allowExit(true);
+                        this.visitList();
+                    }
+                }.bind(this),
+                buttons: {
+                    confirm: {
+                        label:'Discard',
+                        className:'btn-danger'
+                    }
+                }
+            });
+        }
+    },
+    changeMessage: function (text, css, timeout) {
```

Comment:
There is a ViewModel mixin in `changeMessage.js` with convenience methods for displaying messages. You can inherit from that to DRY this up.